### PR TITLE
recursively validate containers

### LINKED ISSUE #1975: Container.validate() doesn't validate sub-containers
**Describe the bug**

Containers contained in Containers do not get validated with the parent container's `.validate()` method is called.

**To Reproduce**
```py3
from ctapipe.core import Container, Field

class Sub(Container):
    y = Field(0*u.m, "Sub.y", unit=u.m)

class A(Container):
    x = Field(0*u.m, "A.x", unit=u.m)
    sub = Field(None, "sub", type=Sub)

# works:
try: 
    A(x=10*u.s, sub=Sub(y=10*u.m)).validate()
except Exception as err:
    print(err)

#> A Field 'x': the value '10.0 s' (<class 'astropy.units.quantity.Quantity'>) is invalid: : 's' (time) and 'm' (length) are not convertible

# doesn't not raise exception (the sub field doesn't get checked recursively):
try: 
    A(x=10*u.m, sub=Sub(y=10*u.s)).validate()
except Exception as err:
    print(err)
```

**Expected behavior**
Should raise exception if any sub-Container also fails validation

### PR DESCRIPTION
**Pull request by kosack**
fixes #1975

### CODE REVIEW

File: ctapipe/core/container.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 from copy import deepcopy
 from pprint import pformat
 from textwrap import wrap
+from typing import Container
```
Comment (maxnoe) [inline ctapipe/core/container.py:5]: Not the container you are looking for

**Review by maxnoe — DISMISSED** [review]

File: ctapipe/core/container.py [hunk 2]
```diff
@@ -108,6 +108,10 @@ def validate(self, value):
                 f"{errorstr} Should be an instance of {self.type}"
             )
 
+        if isinstance(value, Container):
```
[the remark this answers is not in the document]
Comment (maxnoe) [reply ctapipe/core/container.py:111]: No, inside of methods the Container is already defined.

[the remark this answers is not in the document]
Comment (maxnoe) [reply ctapipe/core/container.py:111]: You can of course refer to the class inside of methods...

**Comment by kosack** [comment]
@maxnoe sorry , you can re-review this now - i added also support for Maps, so you can now do `event.validate()` and it checks everything

File: ctapipe/core/container.py [hunk 3]
```diff
@@ -112,6 +112,11 @@ def validate(self, value):
             # recursively check sub-containers
             value.validate()
 
+        if isinstance(value, Map):
+            for key, map_value in value.items():
```
Comment (maxnoe) [inline ctapipe/core/container.py:116]: key is not needed right? Or better, catch the validation error and report the key so users can better see what failed?

Comment (maxnoe) [reply ctapipe/core/container.py:116]:
```
try:
   map_value.validate()
except Exception as e:
   raise FieldValidationError(f"Failed to validate map entry {key}: {e}") from e
```

Comment (kosack) [reply ctapipe/core/container.py:116]: Good idea.  Now the resulting error message is for example:
```
FieldValidationError: ArrayEventContainer Field 'dl1': DL1Container Field 'tel': [3]: DL1CameraContainer Field 'parameters': ImageParametersContainer Field 'hillas': HillasParametersContainer Field 'r': the value '10.0 m' (<class 'astropy.units.quantity.Quantity'>) is invalid: : 'm' (length) and 'deg' (angle) are not convertible
```

**Review by maxnoe — APPROVED** [review]

**Review by maxnoe — APPROVED** [review]

**Comment by maxnoe** [comment]
> sorry , you can re-review this now - i added also support for Maps, so you can now do event.validate() and it checks everything

We should probably add this to a couple of unit tests, e.g. image processor and shower processor tests.

File: ctapipe/core/container.py [hunk 4]
```diff
@@ -108,6 +108,18 @@ def validate(self, value):
                 f"{errorstr} Should be an instance of {self.type}"
             )
 
+        if isinstance(value, Container):
+            # recursively check sub-containers
+            value.validate()
+
+        if isinstance(value, Map):
```
Comment (nbiederbeck) [inline ctapipe/core/container.py:115]: shouldn't this return early? If value is a Map, the other validity checks don't need to run?

same above.

and if yes we may need to create a test that checks this?

Comment (kosack) [reply ctapipe/core/container.py:116]: I added some returns for both cases. No need to check other attributes for Map or Container fields.

**Review by maxnoe — APPROVED** [review]

**Review by nbiederbeck — APPROVED** [review]

**Pull request merged**